Update debugging docs to mbed OS 5

janjongboom opened the pull request:
@iriark01 Work in progress

The idea is to:

* Update these docs to mbed OS 5 and mbed CLI
* Explain other ways than just pyOCD, as it does not work for all boards
* Explain how to make debug builds
* Packages required to set up debugging for any GDB-supporting IDE

It should replace (and I think all of these should redirect to a page in this space when this lands):

* https://developer.mbed.org/handbook/SerialPC
* https://developer.mbed.org/handbook/Windows-serial-configuration
* https://developer.mbed.org/handbook/Debugging
* https://docs.mbed.com/docs/getting-started-mbed-os/en/latest/Full_Guide/Debugging/
* https://developer.mbed.org/users/maclobdell/notebook/debugging-with-pyocd/
* https://developer.mbed.org/cookbook/eclipse-for-building-and-debugging
* https://developer.mbed.org/cookbook/Using-mbed-with-gcc-and-eclipse
* https://developer.mbed.org/users/c1728p9/notebook/debugging-mbed-50-projects-with-eclipse-and-pyocd/

The flow should be (and the docs should guide people through this):

1. Set up serial communication.
2. Set up a debug connection.
3. Use an IDE to debug.

In a way that covers 98% of all boards, and is fool proof.

janjongboom commented:
@iriark01 Alright, it starts to shape up now. Need images etc. but the content should be relatively coherent.

iriark01 commented:
Happy to replace the mbed OS 5 handbook pages when this is ready. For the rest, it's up to Simon F.

If you make something foolproof, someone will make a better fool.

iriark01 commented:
@AnotherButler can you please edit these for Jan? Don't worry about organizing the docs in the YAML and so on. Just work with him on the wording so we can be ready to roll. mbed Cloud is taking up too much time this week and it's not fair to Jan.

AnotherButler commented:
@iriark01 Absolutely. I'd be happy to.

AnotherButler commented on Docs/Debugging/Debugging_Eclipse_pyOCD.md:
```diff
@@ -1,67 +0,0 @@
-#Debugging with Eclipse and pyOCD
```
Query: Why are we deleting this page when we added it less than two weeks ago?

janjongboom replied:
@AnotherButler This is now in eclipse.md

AnotherButler commented on Docs/Debugging/Debugging_Eclipse_pyOCD.md:
```diff
@@ -1,67 +0,0 @@
-#Debugging with Eclipse and pyOCD
-
-Through debugging, you can do things such as set breakpoints, set watchpoints, view registers, view disassembly, browse memory and examine the callstack. This article explains how to debug an exported mbed project with Eclipse and pyOCD. To debug a project, first, import it with mbed CLI, then create an Eclipse Makefile project and finally set up debugger configurations. You can find more detaile ...
```
Because we've deleted this page, there is no longer information about why anyone would want to debug or what one can do with debugging. Please include this information elsewhere. Maybe on the first page?

janjongboom replied:
Done

AnotherButler commented on Docs/Debugging/Keil.md:
```diff
@@ -1,122 +0,0 @@
-#CMSIS DAP MDK
```
It seems like there's some useful information in here that we lose by deleting this page.

janjongboom replied:
@AnotherButler I'm a bit unsure about this. The information is superseded by the info in 'Debugging with uVision' (upgraded to uVision 5), but the old Keil.md looks a bit more information dense, I'm wondering if we need to explain to people what a breakpoint is... Especially if they just shelled out XXXX for a uVision license.

AnotherButler commented on Docs/Debugging/toolchain.md:
```diff
@@ -0,0 +1,105 @@
+# Setting up a local debug toolchain
+
+Most mbed-enabled development boards contain two chips: the target microcontroller and a CMSIS-DAP interface chip. The most common use case of the interface chip is mounting your development board as a USB mass-storage device, making it easy to flash new firmware. However, you can also use this interface chip to debug the target microcontroller without needing an external debugger like a JTAG.
```
Because we deleted the pyOCD page, this page now holds the only information about CMSIS-DAP. Should we include a brief sentence explaining what it is [mbed's debugging protocol]?

AnotherButler commented on Docs/Debugging/offline.md:
```diff
@@ -1,317 +0,0 @@
-
-* Cover ways to set up your local toolchain.
-* Show how to debug applications with your favourite IDE.
-* Explain how to sync online and offline code and libraries.
```
Is this information no longer relevant? If it is still relevant, is it available anywhere else?

janjongboom replied:
1. Should be covered by mbed CLI and by toolchain.md in this repo
2. Is covered by this repo
3. Should be covered by mbed CLI

AnotherButler commented on Docs/Debugging/uvision.md:
```diff
@@ -0,0 +1,36 @@
+
+This document explains how to build and debug mbed OS applications using Keil uVision 5. Due to the linker limits this does not work in the free version of uVision. If you do not have a uVision license you can use [Eclipse](Eclipse.md), [Visual Studio Code](VsCode.md) or any other IDE that supports debugging through GDB ([more info](toolchain.md)).
+
+@todo: which drivers do people need?
```
Query: Do we still need this information? Can I help you find it? If not, please delete this line.

AnotherButler commented on Docs/Debugging/toolchain.md:
````diff
@@ -0,0 +1,105 @@
+```
+$ arm-none-eabi-gdb
+GNU gdb (GNU Tools for ARM Embedded Processors) 7.8.0.20150604-cvs
+Copyright (C) 2014 Free Software Foundation, Inc.
````
Query: Should this be 2014? Please confirm.

janjongboom replied:
Yes. It's literal output from the gdb process.

AnotherButler commented on Docs/Debugging/toolchain.md:
````diff
@@ -0,0 +1,105 @@
+```
+$ pyocd-gdbserver
+
+INFO:root:DAP SWD MODE initialised
````
Query: Should this be "initialized" for U.S. English? @iriark01

janjongboom replied:
@AnotherButler This is output by pyOCD, so if it needs to be changed then @c1728p9 would need to do so.

c1728p9 replied:
I created a [PR in pyOCD](https://github.com/mbedmicro/pyOCD/pull/274) to update the output so you can use the correct spelling here.

AnotherButler commented on Docs/Debugging/printf.md:
```diff
@@ -1,106 +1,231 @@
+This associates the macro's name (`MACRO_NAME`) with the body of the macro, represented by whatever `value` appears between the first space after the `MACRO_NAME` and the end of the line.
 
-``printf()``s are very useful for debugging when looking for an explanation to a problem. Otherwise, it is nice to be able to disable many of them. We can use the ``#define`` directive to create parameterized macros that extend the basic ``printf()`` functionality. For example, macros can expand to printf()s when needed, but to empty statements under other conditions. 
+``printf()`` is very useful for debugging when looking for an explanation to a problem. Otherwise, it is nice to be able to disable many of them. We can use the ``#define`` directive to create parameterized macros that extend the basic ``printf()`` functionality. For example, macros can expand to printf()s when needed, but to empty statements under other conditions.
```
Query: What does "them" refer to in "it is nice to be able to disable many of them"?

AnotherButler commented on Docs/Debugging/Eclipse.md:
```diff
@@ -0,0 +1,101 @@
+### pyOCD
+
+1. Select *Run > Debug Configurations...*.
+1. If no configuration exists under *GDB pyOCD Debugging*, double click it to create a new configuration.
```
Query: What are you double clicking? GDB pyOCD Debugging?

janjongboom commented:
Because I saw some confusion (and not to be penalized by search engines) I renamed eclipse.md and uvision.md back to their old names, so it doesn't look like I deleted some random pages :-)

Also updated images and removed old images.

@AnotherButler

AnotherButler commented:
Thanks, @janjongboom. I will wait for @c1728p9 to complete his review of pyOCD output, etc. before merging.

janjongboom commented:
@AnotherButler Did a run through the docs, see ^

AnotherButler commented:
@janjongboom Thanks for looking through the docs. Could you please split this PR into smaller PRs or squash the commits? I'm starting to have trouble following all of the changes.

Also, are the Lights of the Dead the same as Siren Lights? If not, where did they come from? Are the runtime error, serial terminal debug messages, using the LEDs and debugging incorrect data values examples and snippets deprecated or incorrect?

janjongboom commented:
@AnotherButler Squashed them.

* Lights of the Death are Siren Lights, yes. Chose that name because Google (1,540 vs. 7 results).
* Runtime error - deprecated, we have way better tools to debug this. Same for debugging using LEDs. Feels like this was written before we had serial output.
* Serial terminal debug messages - are all in printf.md.
* Debugging incorrect values, also covered by printf.md.

AnotherButler commented:
@janjongboom Thanks for squashing, and thanks for explaining. I'm sorry, but I still can't seem to see where the serial terminal debug messages example and the debugging incorrect values example are in printf. Could you tell me more specifically where to look?

janjongboom commented:
@AnotherButler 

The serial terminal debugging example is not there literally, but it's covered by [1](https://github.com/janjongboom/Debugging-docs/blob/mbedos5/Docs/Debugging/printf.md#example-program) and [2](https://github.com/janjongboom/Debugging-docs/blob/mbedos5/Docs/Debugging/printf.md#seeing-the-log-messages) in more detail than it is in the current docs.

On 'Debugging incorrect values':

> It is often useful to find out the values of variables in your program at a given time. The example is the same as the one above, except that we are are now using the serial port to print out the value of the AnalogIn. You might need to use this method if you are using a data value to determine the program flow.

This sentence is the premise for printf.md, covered by [the introduction](https://github.com/janjongboom/Debugging-docs/blob/mbedos5/Docs/Debugging/printf.md#debugging-using-printf-statements) and [example program](https://github.com/janjongboom/Debugging-docs/blob/mbedos5/Docs/Debugging/printf.md#example-program). The idea of that section - including the code - was to demonstrate how to show values from a sensor over a serial connection, which (I think) is well covered by printf.md. The specific example for AnalogIn only made sense because the docs were not very good.

Thanks by the way for going over this in great detail :-).

AnotherButler commented:
@janjongboom Ah, that makes sense. Thank you so much for taking the time to explain that to me. I really appreciate it. :)

AnotherButler commented:
@iriark01 These look good to me. Please confirm.
